File: py/candlestick_patterns/patterns/long_line.py
"""Long Line pattern (1-candle)."""
from __future__ import annotations

from ..core.primitives import is_white, real_body, upper_shadow, lower_shadow
from ...fuzzy import t_product_all


def long_line(self) -> float:
    """Long Line: a one-candle pattern.

    Must have:
    - long real body,
    - short upper shadow,
    - short lower shadow.

    The meaning of "long" is specified with self._long_body.
    The meaning of "short" for shadows is specified with self._short_shadow.

    Category B: direction from candle color.

    Returns:
        Continuous float in [-100, +100].
    """
    if not self._enough(1, self._long_body, self._short_shadow):
        return 0.0

    o, h, l, c = self._bar(1)

    # Fuzzy: long body, short shadows.
    mu_long = self._mu_greater(real_body(o, c), self._long_body, 1)
    mu_us = self._mu_less(upper_shadow(o, h, c), self._short_shadow, 1)
    mu_ls = self._mu_less(lower_shadow(o, l, c), self._short_shadow, 1)

    confidence = t_product_all(mu_long, mu_us, mu_ls)

    # Crisp direction from color.
    direction = 1 if is_white(o, c) else -1

    return direction * confidence * 100.0

File: ts/candlestick-patterns/patterns/long-line.ts
/** Long Line pattern (1-candle). */
import { CandlestickPatternsEngine } from '../core/engine.ts';
import { tProductAll } from '../../fuzzy/index.ts';
import { realBody, upperShadow, lowerShadow, isWhite } from '../core/primitives.ts';

/**
 * Long Line: a one-candle pattern.
 *
 * Must have:
 * - long real body,
 * - short upper shadow,
 * - short lower shadow.
 *
 * The meaning of "long" is specified with `cp.longBody`.
 * The meaning of "short" for shadows is specified with `cp.shortShadow`.
 *
 * Category B: direction from candle color.
 *
 * Returns:
 *     Continuous float in [-100, +100].
 */
export function longLine(cp: CandlestickPatternsEngine): number {
    if (!cp.enough(1, cp.longBody, cp.shortShadow)) return 0.0;

    const b = cp.bar(1);
    // Fuzzy: long body, short shadows.
    const muLong = cp.muGreaterCS(realBody(b.o, b.c), cp.longBody, 1);
    const muUS = cp.muLessCS(upperShadow(b.o, b.h, b.c), cp.shortShadow, 1);
    const muLS = cp.muLessCS(lowerShadow(b.o, b.l, b.c), cp.shortShadow, 1);

    const confidence = tProductAll(muLong, muUS, muLS);
    // Crisp direction from color.
    const direction = isWhite(b.o, b.c) ? 1 : -1;
    return direction * confidence * 100.0;
}
